File: services/main-api/internal/service/search_service.go
package service

import (
	"fmt"
	"strings"

	"go-stream/services/main-api/internal/domain"
	"gorm.io/gorm"
)

type dbSearchService struct {
	db *gorm.DB
}

func NewSearchService(db *gorm.DB) domain.SearchService {
	return &dbSearchService{db: db}
}

// SearchGlobal queries the database directly using ILIKE.
// This is a temporary solution before moving to a dedicated search engine.
func (s *dbSearchService) SearchGlobal(query string, limit int) (*domain.SearchResult, error) {
	if query == "" {
		return &domain.SearchResult{
			Rooms:   []domain.Room{},
			Authors: []domain.Author{},
			Games:   []domain.Game{},
		}, nil
	}

	// PostgreSQL uses ILIKE for case-insensitive search.
	likeQuery := fmt.Sprintf("%%%s%%", strings.TrimSpace(query))
	if limit <= 0 || limit > 50 {
		limit = 10
	}

	result := &domain.SearchResult{
		Rooms:   make([]domain.Room, 0),
		Authors: make([]domain.Author, 0),
		Games:   make([]domain.Game, 0),
	}

	// 1. Search Live Rooms (only live and public)
	err := s.db.Preload("Host").Preload("Category").Preload("Game").Preload("Tags").
		Joins("LEFT JOIN room_tags ON room_tags.room_id = rooms.id").
		Joins("LEFT JOIN tags ON tags.id = room_tags.tag_id").
		Where("rooms.status = ? AND rooms.visibility = ?", domain.RoomStatusLive, domain.RoomVisibilityPublic).
		Where("(rooms.title ILIKE ? OR rooms.description ILIKE ? OR tags.name ILIKE ?)", likeQuery, likeQuery, likeQuery).
		Group("rooms.id").
		Limit(limit).
		Find(&result.Rooms).Error
	if err != nil {
		return nil, fmt.Errorf("failed to search rooms: %w", err)
	}

	// 2. Search Authors
	// Need to join User to search by Username or DisplayName
	err = s.db.Preload("User").
		Joins("JOIN users ON authors.user_id = users.id").
		Where("authors.display_name ILIKE ? OR users.name ILIKE ? OR authors.bio ILIKE ?", likeQuery, likeQuery, likeQuery).
		Limit(limit).
		Find(&result.Authors).Error
	if err != nil {
		return nil, fmt.Errorf("failed to search authors: %w", err)
	}

	// 3. Search Games
	err = s.db.Preload("Category").
		Where("name ILIKE ? OR slug ILIKE ?", likeQuery, likeQuery).
		Limit(limit).
		Find(&result.Games).Error
	if err != nil {
		return nil, fmt.Errorf("failed to search games: %w", err)
	}

	return result, nil
}

File: fe/src/services/search.service.ts
const BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";

export interface SearchRoom {
  id: number;
  title: string;
  thumbnail?: string;
  viewer_count: number;
  host: { name: string; slug: string };
}

export interface SearchAuthor {
  id: number;
  display_name: string;
  bio?: string;
  avatar?: string;
  user: { name: string; slug?: string };
}

export interface SearchGame {
  id: number;
  name: string;
  slug: string;
  cover_image?: string;
  category?: { name: string; slug: string };
}

export interface SearchResults {
  rooms: SearchRoom[];
  authors: SearchAuthor[];
  games: SearchGame[];
}

export async function searchGlobal(query: string): Promise<SearchResults | null> {
  try {
    const response = await fetch(`${BASE_URL}/api/search?q=${encodeURIComponent(query)}&limit=12`, {
      cache: "no-store",
    });
    if (!response.ok) return null;
    const result = await response.json();
    return (result.data ?? result) as SearchResults;
  } catch {
    return null;
  }
}
